### PR TITLE
T13: Update MultiSelection when deleting open Thread

### PR DESCRIPTION
**Pull request by KevinBoulongne**
Depends on #1584

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt [hunk 1]
```diff
@@ -524,6 +524,16 @@ class ThreadListFragment : TwoPaneFragment(), SwipeRefreshLayout.OnRefreshListen
 
             waitingBeforeNotifyAdapter = threadListViewModel.isRecoveringFinished
 
+            deletedItemsIndices = { indices ->
+                if (isMultiSelectOn) {
+                    val previousThreads = threadListAdapter.dataSet.filterIsInstance<Thread>()
+                    indices.forEach {
+                        val isRemoved = mainViewModel.selectedThreads.remove(previousThreads[it])
+                        if (isRemoved) publishSelectedItems()
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt:532]: Publish after the forEach

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt [hunk 2]
```diff
@@ -525,6 +525,17 @@ class ThreadListFragment : TwoPaneFragment(), SwipeRefreshLayout.OnRefreshListen
...
             waitingBeforeNotifyAdapter = threadListViewModel.isRecoveringFinished
 
+            deletedItemsIndices = { indices ->
+                if (isMultiSelectOn) {
+                    val previousThreads = threadListAdapter.dataSet.filterIsInstance<Thread>()
+                    var isRemoved = false
+                    indices.forEach {
+                        isRemoved = mainViewModel.selectedThreads.remove(previousThreads[it])
+                    }
+                    if (isRemoved) publishSelectedItems()
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt:535]: remove() only returns false if the object wasn't in the data list, but with our code here, if only one of the selected thread isn't in the selection, then we don't publish and all the other selected thread won't be deleted in the UI

File: app/src/main/java/com/infomaniak/mail/utils/RealmChangesBinding.kt [hunk 3]
```diff
@@ -87,8 +88,10 @@ class RealmChangesBinding<T : BaseRealmObject, VH : ViewHolder> private construc
 
             is UpdatedResults -> { // Any update
                 waitingBeforeNotifyAdapter?.observeWaiting {
+                    deletedItemsIndices?.invoke(resultsChange.deletions)
                     notifyAdapter()
                 } ?: run {
+                    deletedItemsIndices?.invoke(resultsChange.deletions)
                     notifyAdapter()
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/utils/RealmChangesBinding.kt:95]: We could factorize in a function now

File: app/src/main/java/com/infomaniak/mail/utils/RealmChangesBinding.kt [hunk 4]
```diff
@@ -115,8 +118,10 @@ class RealmChangesBinding<T : BaseRealmObject, VH : ViewHolder> private construc
 
             is UpdatedList -> { // Any update
                 waitingBeforeNotifyAdapter?.observeWaiting {
+                    deletedItemsIndices?.invoke(listChange.deletions)
                     notifyAdapter()
                 } ?: run {
+                    deletedItemsIndices?.invoke(listChange.deletions)
                     notifyAdapter()
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/utils/RealmChangesBinding.kt:125]: We could factorize in a function now

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt (hunk 2 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt]: True. It's fixed.

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt [hunk 5]
```diff
@@ -521,6 +521,18 @@ class ThreadListFragment : TwoPaneFragment(), SwipeRefreshLayout.OnRefreshListen
...
+                if (isMultiSelectOn) {
+                    val previousThreads = threadListAdapter.dataSet.filterIsInstance<Thread>()
+                    var shouldPublish = false
+                    indices.forEach {
+                        val isRemoved = mainViewModel.selectedThreads.remove(previousThreads[it])
+                        if (isRemoved) shouldPublish = true
+                    }
+                    if (shouldPublish) publishSelectedItems()
+                }
+            }
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt:534]: You should extract this in its own function 
That would solve the sonar code smell also

**Review by FabianDevel — CHANGES_REQUESTED** [review]

**Review by LunarX — APPROVED** [review]

**Review by FabianDevel — APPROVED** [review]

**Pull request merged**